### ISSUE TITLE
Tough time running Stan in Julia

### ISSUE THREAD

**Issue by gregbritten** [jpmattern/bayesian_cbiomes#1]
Wu (lab mate) and I tried installing and running Stan in Julia yesterday and had a bitch of a time. It's still not working on my PC. Wu was able to get it to work on his Mac after having to fix several weird paths and settings. (I remember folks in the past having a harder time with Julia). Could you give it a go on your system and let me know if you run into issues? We'll have to sort out potential issues if we're going to support Julia in the workshop.. 
I'm going to give it another go on my PC and will post a question to the author of CmdStan.jl/Stan.jl
I'm also going to try it on my old Mac to see how it goes.

**Comment by jpmattern** [comment]
OK, I'll try a bit more as well...

BTW, I noticed that you are writing the installation instructions based on the OS first. I'll follow that example and start writing instructions for Linux...

**Comment by AWO-code** [comment]
I am trying to install and run Stan in Julia right now. I ran into the JULIA_CMDSTAN_HOME issue. So how do you set JULIA_CMDSTAN_HOME?

**Comment by gregbritten** [comment]
This will probably fix you up:
https://github.com/StanJulia/CmdStan.jl/blob/master/docs/src/INSTALLATION.md

We will add these steps to the install instructions...

**Comment by AWO-code** [comment]
OK, thanks! Now, I am getting the following scary error message:

ERROR: error when compiling Stan model at /var/folders/bm/6jw569yj77l9l56ql9s5lrth0000gn/T/jl_0xFUWG/mod.stan
    (CmdStan home: /Users/omta/Julia/Stan/cmdstan):
fatal error: malformed or corrupted AST file: 'could not find file '/Users/omta/Stan/cmdstan/stan/lib/stan_math/lib/boost_1.69.0/boost/exception/detail/clone_current_exception.hpp' referenced by AST file 'stan/src/stan/model/model_header.hpp.gch''
1 error generated.
make: *** [/var/folders/bm/6jw569yj77l9l56ql9s5lrth0000gn/T/jl_0xFUWG/mod] Error 1

**Comment by gregbritten** [comment]
Yikes. Did a bit of Googling and it seems that 'malformed or corrupted AST file' is associated with a failure of your Xcode C++ compiler. (Oddly I did not find anyone reporting this error using Stan). Did you update Xcode or iOS recently? If not, perhaps try updating Xcode?
Maybe @gaelforget has some insight on how Julia interacts with Xcode in iOS?

**Comment by AndrewIrwin** [comment]
I've tried a new installation using the steps on this page for stan + Julia on OSX / MacOS. Similar problem as above
ERROR: The environment variable JULIA_CMDSTAN_HOME needs to be set.
ERROR: UndefVarError: mod_compile not defined
Adding instructions for setting the environment variable (from within Julia, or just in the shell?) to the instructions would be helpful.

**Comment by AndrewIrwin** [comment]
All good after adding
JULIA_CMDSTAN_HOME=~/cmdstan
to .bash_profile

**Comment by maxEntropyProd** [comment]
I had no problem running STAN under python via Jupyter under Windows, but I was unable to even build cmdstan for "StanSample.jl".  I haven't used Anaconda on Windows (or anywhere), but it does not seem to install basic linux utils.  I used conda to install _make_ from anaconda channel, but it complains about not finding /bin/sh when attempting to execute _make build_, so it's expecting at least some minimal bash shell.  

I uninstalled the anaconda make and installed one under the conda-forge channel.  It got a bit farther in some ways but failed issuing the following:

```
(base) PS C:\users\jvallino\stan\cmdstan> make build
'cut' is not recognized as an internal or external command,
operable program or batch file.
'cut' is not recognized as an internal or external command,
operable program or batch file.
File not found - *cvodes*.hpp
File not found - *idas*.hpp
File not found - *kinsol*.hpp
File not found - *_inst.cpp
File not found - *_def.cpp
File not found - *.hpp
g++ -std=c++1y -m64 -D_REENTRANT -Wall -Wno-unused-function -Wno-uninitialized -Wno-unused-but-set-variable -Wno-unused-variable -Wno-sign-compare -Wno-unused-local-typedefs      -I stan/lib/stan_math/lib/tbb_2019_U8/include -O3 -I src -I stan/src -I lib/rapidjson_1.1.0/ -I stan/lib/stan_math/ -I stan/lib/stan_math/lib/eigen_3.3.3 -I stan/lib/stan_math/lib/boost_1.72.0 -I stan/lib/stan_math/lib/sundials_4.1.0/include  -D_USE_MATH_DEFINES  -DBOOST_DISABLE_ASSERTS      -c -MT src/cmdstan/stanc.o -M -E -MG -MP -MF src/cmdstan/stanc.d src/cmdstan/stanc.cpp
'cut' is not recognized as an internal or external command,
operable program or batch file.
'cut' is not recognized as an internal or external command,
operable program or batch file.
File not found - *cvodes*.hpp
File not found - *idas*.hpp
File not found - *kinsol*.hpp
File not found - *_inst.cpp
File not found - *_def.cpp
File not found - *.hpp
The syntax of the command is incorrect.
make: *** [make/stanc:29: bin/stanc.exe] Error 1
```

g++ seems to run OK, but I don't see any binary produced. I also don't understand the complaint about 'cut' from make.  It also looks like it's having problems finding headers for sundials, given cvoids.hpp and similar files not being found. I will continue to mess around (at least for a bit), but any ideas are appreciated.

**Comment by gaelforget** [comment]
Example worked fine for me too (once `JULIA_CMDSTAN_HOME` / `CMDSTAN_HOME` is set) except that for some reason I cannot update `DataFrames.jl` to its `v0.20.0` and the `read_summary` call returns an error with `v0.19.4` of `DataFrames.jl`.

```ERROR: function rename! does not accept keyword arguments```

In case others are in the same situation, here is a work-around : 

```
using CSV, DataFrames

function read_summary_gf(m::T) where {T <: CmdStanModels}
         fname = "$(m.output_base)_summary.csv"
         !isdir(fname) && stan_summary(m)

         df = CSV.read(fname, delim=",", comment="#")

         bnames=names(df)
         cnames = lowercase.(convert.(String, String.(bnames)))
         cnames[1] = "parameters"
         cnames[4] = "std"
         cnames[8] = "ess"
         tmp=Dict(bnames[i] => Symbol(cnames[i]) for i=1:length(bnames))
         rename!(df, tmp)

         df[!, :parameters] = Symbol.(df[!, :parameters])

         ChainDataFrame("CmdStan Summary", df)
end   # end of read_samples

println(read_summary_gf(mod_compile));
```

**Comment by maxEntropyProd** [comment]
I thought I was making progress

1. I was able to compile cmdstan under Ubuntu under Windows 10 (but had to still install make and g++)

1. I had to also install CSV, DataFrames, etc packages, but that worked fine.

1. Had the env problems, but under Jupyter, I added ENV["JULIA_CMDSTAN_HOME"]="C:\\Users\\jvallino\\GitHub\\bayesian_cbiomes\\cmdstan" etc, 

1. But when I got to: mod = SampleModel("linear_reg", stancode), I still get this same strange error:

````
C:\Users\jvallino\AppData\Local\Temp\jl_QoN1gF\linear_reg.stan updated.
error when compiling Stan model at C:\Users\jvallino\AppData\Local\Temp\jl_QoN1gF\linear_reg.stan
    (CmdStan home: C:\Users\jvallino\GitHub\bayesian_cbiomes\cmdstan):
'cut' is not recognized as an internal or external command,
operable program or batch file.
'cut' is not recognized as an internal or external command,
operable program or batch file.
File not found - *cvodes*.hpp
File not found - *idas*.hpp
File not found - *kinsol*.hpp
File not found - *_inst.cpp
File not found - *_def.cpp
File not found - *.hpp
The syntax of the command is incorrect.
make: *** [make/stanc:29: bin/stanc.exe] Error 1


Stacktrace:
 [1] ensure_executable(::StanModel{String}) at C:\Users\jvallino\.julia\packages\StanRun\E3YcJ\src\StanRun.jl:123
 [2] #SampleModel#6(::StanSample.Sample, ::StanBase.RandomSeed, ::StanBase.Init, ::StanBase.Output, ::String, ::Type{SampleModel}, ::String, ::String, ::Array{Int64,1}) at C:\Users\jvallino\.julia\packages\StanSample\e3hsD\src\stanmodel\SampleModel.jl:57
 [3] SampleModel at C:\Users\jvallino\.julia\packages\StanSample\e3hsD\src\stanmodel\SampleModel.jl:51 [inlined] (repeats 2 times)
 [4] top-level scope at In[11]:1
````
Seems like there is something I need to add under the Anaconda environment (like the set of linux tools)

**Comment by gregbritten** [comment]
Joe, haven't seen this error pop up yet.. Unfortunately the connection between Julia and cmdstan has been producing some incomprehensible errors. If issues persist I recommend sticking with your working PyStan implementation.

**Comment by maxEntropyProd** [comment]
Right.  I'm now trying to go entirely via the Linux windows subsystem
instead of trying to get windows and Linux to play well together. But, just
getting off train so had to stop, but will pick up at meeting

Thanks!
